File: src/bindings/viewport.cpp
#include "bindings.hpp"

#include "../viewport.hpp"
// Include required because viewport.hpp forward declares Document.
#include "../document.hpp" // IWYU pragma: keep.
// Include required because viewport.hpp forward declares DocumentView.
#include "../document_view.hpp" // IWYU pragma: keep.

void ViewportBinding::init_bridge(sol::table& core) {
    // clang-format off
    core.new_usertype<Viewport>("Viewport",
        /* Properties. */
        "view", sol::readonly(&Viewport::view_),

        /* Functions. */
        "change_document_view", &Viewport::change_document_view,
        "adjust", &Viewport::adjust_viewport,
        "scroll_up", [](Viewport& self, const std::size_t n) -> void { self.scroll_up(n); },
        "scroll_down", [](Viewport& self, const std::size_t n) -> void { self.scroll_down(n); },
        "scroll_left", [](Viewport& self, const std::size_t n) -> void { self.scroll_left(n); },
        "scroll_right", [](Viewport& self, const std::size_t n) -> void { self.scroll_right(n); });
    // clang-format on
}

File: lua/bindings/viewport.lua
--- @meta

--- Viewports abstract Display regions occupied by a Window. They draw a Document and optional gutter and mode line.
--- @class Core.Viewport
--- @field view Core.DocumentView View of the to displayable Document
Core.Viewport = {}

--- Changes the displayed DocumentView.
--- @param view Core.DocumentView
function Core.Viewport:change_document_view(view) end

--- Adjusts the Viewport to correctly display the cursor.
function Core.Viewport:adjust() end

--- Moves the Viewport up.
--- @param n integer
function Core.Viewport:scroll_up(n) end

--- Moves the Viewport down.
--- @param n integer
function Core.Viewport:scroll_down(n) end

--- Moves the Viewport to the left.
--- @param n integer
function Core.Viewport:scroll_left(n) end

--- Moves the Viewport to the right.
--- @param n integer
function Core.Viewport:scroll_right(n) end
